FineModuleTracing: only collect fine-grained tracing for those source files that compiler promises to type check.

nkcsgexi opened the pull request:
rdar://141065994

nkcsgexi committed:
FineModuleTracing: only collect fine-grained tracing for those source files that compiler promises to type check.

rdar://141065994

tshortli commented on lib/FrontendTool/LoadedModuleTrace.cpp:
```diff
@@ -915,8 +899,9 @@ static void createFineModuleTraceFile(const InputFile &input, ModuleDecl *MD) {
+  ModuleDecl *mainModule = Instance.getMainModule();
   // When lazy type checking is enabled, we may end up with a partial AST.
   // Walking on these partial AST completely may expose latent bugs.
   if (mainModule->getASTContext().TypeCheckerOpts.EnableLazyTypecheck)
```
Could you also switch this to skipping emitting the trace for emit module actions, instead of querying lazy typechecking?

nkcsgexi replied:
Happy to make that change! Is there a canonical flag I can use from here?

tshortli replied:
A `switch` over the `FrontendOptions::ActionType` for the job is probably the best way to comprehensively handle this.

tshortli replied:
`FrontendOpts.RequestedAction`

nkcsgexi committed:
FineModuleTracing: use requested action type to further gate whether emitting fine-grained module trace is supported.

Emitting fine-grained module tracing files during full compilation should be sufficient for the use case in design. Other
swift-frontend invocations, such as emitting modules, doesn't need fully type-checked AST to perform the action, therefore
requesting/walking the fully typechecked AST to emit those tracing files may introduce performance overhead.

nkcsgexi commented:
@swift-ci please smoke test

nkcsgexi commented:
@swift-ci please smoke test linux

nkcsgexi commented:
also cc: @slavapestov

nkcsgexi commented:
@swift-ci please smoke test linux
